sl_server.cc

```diff
@@ -0,0 +1,353 @@
+  Users users;
+  users.ParseFromString(users_serial);
+  for(int i = 0; i < users.username_size(); i++) {
+    if(users.username(i) == request->username()) {
+      valid_user = true;
+      break;
+    }
+  }
+  if(!valid_user) {
+    mtx_.unlock();
+    return Status(StatusCode::INVALID_ARGUMENT, "user does not exist");
+  }
+
+  // get chirps from database
+  std::string chirps_serial = client_.get("chirps");
+  Chirps chirps;
+  chirps.ParseFromString(chirps_serial);
+
+  // check that parent_id exists in database
+  try {
+    if( (request->parent_id() != "-1") && ((std::stoi(request->parent_id()) >= chirps.chirps_size()) || (std::stoi(request->parent_id()) < 0)) ) {
+      mtx_.unlock();
+      return Status(StatusCode::INVALID_ARGUMENT, "parent_id not valid");
+    }
+  } catch (std::invalid_argument&) {
+    mtx_.unlock();
+    return Status(StatusCode::INVALID_ARGUMENT, "parent_id not valid");
+  }
+
+  // create new chirp
+  Chirp *chirp = new Chirp();
+
+  // create timestamp for chirp
+  int64_t seconds = google::protobuf::util::TimeUtil::TimestampToSeconds(google::protobuf::util::TimeUtil::GetCurrentTime());
+  int64_t useconds = google::protobuf::util::TimeUtil::TimestampToMicroseconds(google::protobuf::util::TimeUtil::GetCurrentTime());
+  chirp::Timestamp* ts = new chirp::Timestamp();
+  ts->set_seconds(seconds);
+  ts->set_useconds(useconds);
+
+  // set variables of chirp
+  chirp->set_allocated_timestamp(ts);
+  chirp->set_username(request->username());
+  chirp->set_text(request->text());
+  chirp->set_id(std::to_string(chirps.chirps_size()));
+  chirp->set_parent_id(request->parent_id());
+  reply->set_allocated_chirp(chirp);
+
+  // add new chirp to chirps repeated variable
+  Chirp *newChirp = chirps.add_chirps();
+  *newChirp = *chirp;
+
+  // add chirp to KVS
+  chirps.SerializeToString(&chirps_serial);
+  client_.put("chirps", chirps_serial);
+
+  std::string replies_serial = client_.get("replies");
+  Replies replies;
+  replies.ParseFromString(replies_serial);
+  Reply* chirpReply = new Reply();
+  chirpReply->set_id(std::stoi(chirp->id()));
+  Reply* newReply = replies.add_allreplies();
+  *newReply = *chirpReply;
+
+  // add reply to original chirp id if reply flag specified
+  if((request->parent_id() != "-1")) {
+    replies.mutable_allreplies(std::stoi(request->parent_id()))
+        ->add_replies(std::stoi(chirp->id()));
+  }
+
+  replies.SerializeToString(&replies_serial);
+  client_.put("replies", replies_serial);
+
+  // broadcast chirp to followers
+  std::string followers_serial = client_.get("followers");
+  Followers followers;
+  followers.ParseFromString(followers_serial);
+
+  // loop through each follower
+  for(int i = 0; i < followers.followers_size(); i++) {
+    if((followers.followers(i)).username() == request->username()) {
+      for(int j = 0; j < followers.followers(i).follows_size(); j++) {
+        std::string follower_name =
+            "monitor::" + followers.followers(i).follows(j);
```

Comment:
Yeah, I fixed it so that my followers are stored as follower::username as the key, and the ids of the followers of the values. So no more complicated levels.